File: Nov2020/21Nov/STPAR-StreetParade.cpp
/*
Problem name: STPAR - Street Parade
Problem link: https://www.spoj.com/problems/STPAR/
*/

#include <bits/stdc++.h>
using namespace std;

int main() {
	ios::sync_with_stdio(0);    
    cin.tie(0);
    string ans="";
	while(true){
    	int n;
    	cin>>n;
    	if(n==0)    
    	    break;
    	int A[n];
    	for(int i=0;i<n;++i){
    		cin>>A[i];
    	}
    	stack<int> s;
    	
    	int count=1;
    	for(int i=0;i<n;++i){
    		if(A[i]==count || (s.size()>0 && s.top()==count)){
    			if((s.size()>0) && (s.top()==count)){
    				s.pop();
    				i--;
    			}	
    			count++;
    			continue;
    		}
    		s.push(A[i]);
    	}
    	while(count<n+1){
    	    if(s.top()!=count){
    	        break;
    	    }
    	    count++;
    	    s.pop();
    	}
    	
    	if(s.size()>0)	ans+="no\n";
    	else	ans+="yes\n";
	}
	cout<<ans;
	return 0;
}

File: Nov2020/21Nov/STPAR-StreetParade.java
/*
Problem Name: STPAR - Street Parade
Problem Link: https://www.spoj.com/problems/STPAR/
*/
import java.util.*;
import java.lang.*;
class Main{
	public static void main (String[] args) throws java.lang.Exception{
		Scanner in = new Scanner(System.in);
		for(;;){
    		int n = in.nextInt();
    		if(n == 0)
    		    break;
    		int arr[] = new int[n];
    		for(int i=0;i<n;i++)
    		    arr[i] = in.nextInt();
    		
    		Stack<Integer> st = new Stack<Integer>();
    		int expected = 1;
    		for(int i=0, ind=0;ind<n;i++){
    		    if((!st.isEmpty()) && (st.peek() == expected)){
    		        st.pop();
    		        expected++;
    		    }
    		    else if(arr[ind] == expected){
    		        ind++;
    		        expected++;
    		    }
    		    else{
    		        st.push(arr[ind++]);
    		    }
    		}
    		boolean possible = true;
    		while(!st.isEmpty()){
    		    if(st.pop() == expected)
    		        expected++;
    		    else{
    		        possible = false;
    		        break;
    		    }
    		}
    		if(possible)
    		    System.out.println("yes");
    		else
    		    System.out.println("no");
		}
	}
}
